models/src/main/scala/coop/rchain/models/rholangn/parmanager/Serialization.scala

```diff
@@ -0,0 +1,581 @@
+        val p2 = readPar()
+        EMinusN(p1, p2)
+
+      case EMULT =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EMultN(p1, p2)
+
+      case EDIV =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EDivN(p1, p2)
+
+      case EMOD =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EModN(p1, p2)
+
+      case ELT =>
+        val p1 = readPar()
+        val p2 = readPar()
+        ELtN(p1, p2)
+
+      case ELTE =>
+        val p1 = readPar()
+        val p2 = readPar()
+        ELteN(p1, p2)
+
+      case EGT =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EGtN(p1, p2)
+
+      case EGTE =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EGteN(p1, p2)
+
+      case EEQ =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EEqN(p1, p2)
+
+      case ENEQ =>
+        val p1 = readPar()
+        val p2 = readPar()
+        ENeqN(p1, p2)
+
+      case EAND =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EAndN(p1, p2)
+
+      case ESHORTAND =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EShortAndN(p1, p2)
+
+      case EOR =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EOrN(p1, p2)
+
+      case ESHORTOR =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EShortOrN(p1, p2)
+
+      case EPLUSPLUS =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EPlusPlusN(p1, p2)
+
+      case EMINUSMINUS =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EMinusMinusN(p1, p2)
+
+      case EPERCENT =>
+        val p1 = readPar()
+        val p2 = readPar()
+        EPercentPercentN(p1, p2)
+
+      case EMETHOD =>
+        val methodName = readString()
+        val target     = readPar()
+        val arguments  = readPars()
+        EMethodN(methodName, target, arguments)
+
+      case EMATCHES =>
+        val target  = readPar()
+        val pattern = readPar()
+        EMatchesN(target, pattern)
+
+      /** Connective */
+      case CONNECTIVE_BOOL      => ConnBoolN
+      case CONNECTIVE_INT       => ConnIntN
+      case CONNECTIVE_BIG_INT   => ConnBigIntN
+      case CONNECTIVE_STRING    => ConnStringN
+      case CONNECTIVE_URI       => ConnUriN
+      case CONNECTIVE_BYTEARRAY => ConnByteArrayN
+
+      case CONNECTIVE_NOT =>
+        val p = readPar()
+        ConnNotN(p)
+
+      case CONNECTIVE_AND =>
+        val ps = readPars()
+        ConnAndN(ps)
+
+      case CONNECTIVE_OR =>
+        val ps = readPars()
+        ConnOrN(ps)
+
+      case CONNECTIVE_VARREF =>
+        val index = readInt()
+        val depth = readInt()
+        ConnVarRefN(index, depth)
+
+      /** Other types */
+      case BUNDLE =>
+        val body      = readPar()
+        val writeFlag = readBool()
+        val readFlag  = readBool()
+        BundleN(body, writeFlag, readFlag)
+
+      case _ => throw new Exception("Invalid tag for ParN deserialization")
```

Comment:
For new core types catch all case should not be used. We know all possible types and matching should enumerate all.

This can be fixed later.